src/team_lib/mhba/xccl_mhba_team.c

```diff
@@ -87,6 +88,56 @@ static void build_rank_map(xccl_mhba_team_t *mhba_team)
     free(data);
 }
 
+static ucs_status_t rcache_reg_mr(void *context, ucs_rcache_t *rcache,void *arg, ucs_rcache_region_t *rregion,
+                                  uint16_t flags){
+    xccl_mhba_team_t *team    = (xccl_mhba_team_t*)context;
+    void *addr                = (void*)rregion->super.start;
+    size_t length             = (size_t)(rregion->super.end - rregion->super.start);
+    xccl_mhba_reg_t* mhba_reg = xccl_rcache_ucs_get_reg_data(rregion);
+    mhba_reg->region = rregion;
+    int* mem_flags = (int*) arg;
+    mhba_reg->mr = ibv_reg_mr(team->node.shared_pd, addr, length, *mem_flags);
+    if (!mhba_reg->mr) {
+        xccl_mhba_error("Failed to register memory");
+        return UCS_ERR_NO_MESSAGE;
+    }
+    return UCS_OK;
+}
+
+static void rcache_dereg_mr(void *context, ucs_rcache_t *rcache, ucs_rcache_region_t *rregion) {
+    xccl_mhba_reg_t* mhba_reg = xccl_rcache_ucs_get_reg_data(rregion);
+    assert(mhba_reg->region == rregion);
+    ibv_dereg_mr(mhba_reg->mr);
+    mhba_reg->mr = NULL;
+}
+
+static xccl_status_t create_rcache(xccl_mhba_team_t* mhba_team) {
+    static ucs_rcache_ops_t rcache_ucs_ops = {
+            .mem_reg     = rcache_reg_mr,
+            .mem_dereg   = rcache_dereg_mr,
+            .dump_region = NULL
+    };
+
+    ucs_rcache_params_t rcache_params;
+    rcache_params.region_struct_size = sizeof(ucs_rcache_region_t)+sizeof(xccl_mhba_reg_t);
+    rcache_params.alignment          = UCS_PGT_ADDR_ALIGN;
+    rcache_params.max_alignment      = ucs_get_page_size();
+    rcache_params.ucm_events         = UCM_EVENT_VM_UNMAPPED |
+                                       UCM_EVENT_MEM_TYPE_FREE;
+    rcache_params.ucm_event_priority = 1000;
+    rcache_params.context            = (void*)mhba_team; //todo maybe change, needed for shared_pd
+    rcache_params.ops                = &rcache_ucs_ops;
+
+    ucs_status_t status = ucs_rcache_create(&rcache_params, "reg cache",
+                                            ucs_stats_get_root(), &mhba_team->context->rcache);
```

Comment:
IF we create 2 teams they both will try to setup rcache twice. I vote for moving this to the context creation - it is a clean solution.